### DISCUSSION TITLE
Build instructions and issues

### DISCUSSION THREAD

**Discussion by maloo** [dotnet/runtime#37833] [General] [answered]
Hi,
just cloned the new runtime repo and tried building by running build.cmd in PS. But I got the error below. Where can I find instructions on how to build it? Could not find a link to any instructions in the readme.
Second, is it still necessary to run build.cmd before using one of the sln files to build a specific library?

```
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
  Determining projects to restore...
  Tool 'coverlet.console' (version '1.7.2') was restored. Available commands: coverlet
  Tool 'dotnet-reportgenerator-globaltool' (version '4.5.8') was restored. Available commands: reportgenerator
  Tool 'microsoft.dotnet.xharness.cli' (version '1.0.0-prerelease.20303.2') was restored. Available commands: xharness

  Restore was successful.
  Restored C:\Users\marcu\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20309.1\tools\Tools.proj (in 25,62 sec).
C:\Users\marcu\.nuget\packages\microsoft.build.traversal\2.0.34\Sdk\Sdk.props(25,18): error MSB4086: A numeric comparison was attempted on "$(TraversalProjectNames.IndexOf($(MSBuildProjectFile), System.StringComparison.OrdinalIgnoreCase))" that evaluates to "-1" instead of a number, in condition " '$(IsTraversal)' == '' And $(TraversalProjectNames.IndexOf($(MSBuildProjectFile), System.StringComparison.OrdinalIgnoreCase)) >= 0 ". [C:\Data\Projects\GitHub\runtime\Build.proj]

Build FAILED.

C:\Users\marcu\.nuget\packages\microsoft.build.traversal\2.0.34\Sdk\Sdk.props(25,18): error MSB4086: A numeric comparison was attempted on "$(TraversalProjectNames.IndexOf($(MSBuildProjectFile), System.StringComparison.OrdinalIgnoreCase))" that evaluates to "-1" instead of a number, in condition " '$(IsTraversal)' == '' And $(TraversalProjectNames.IndexOf($(MSBuildProjectFile), System.StringComparison.OrdinalIgnoreCase)) >= 0 ". [C:\Data\Projects\GitHub\runtime\Build.proj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:45.61
Build failed.
Some builds failed:
        Configuration: Debug, Architecture: x64
```

**Comment by eerhardt** [comment, marked as answer]
Checkout https://github.com/dotnet/runtime/tree/master/docs/workflow. This is linked in the README, under "How can I contribute?"

> Workflow Instructions explains how to build and test

> Second, is it still necessary to run build.cmd before using one of the sln files to build a specific library?

For any sln files under the `src\libraries` folder, yes you need to build.cmd first. See https://github.com/dotnet/runtime/blob/master/docs/workflow/building/libraries/README.md.

**Comment by maloo** [comment]
Thanks, but instruction seems to say the same, run build.cmd (with some parameters if defaults not ok). But whatever I try I get the same error. Tried master and preview5 tag. I have preview5 SDK installed, VS2019 preview. Are there any prerequisites that are not listed to enable build.cmd to run?

**Reply by eerhardt** [reply]
@ViktorHofer @safern - any thoughts?

**Reply by safern** [reply]
I can't see anything obvious from top of my head... This error is coming from the traversal SDK. @rainersigwald @jeffkl do you see anything obvious here?

**Reply by rainersigwald** [reply]
Nothing jumps out at me.

**Reply by maloo** [reply]
How does msbuild handle numbers? like javascript with implicit conversion between strings and numbers or more strict? I created one Condition using x.IndexOf(y) >= 0 which failed if IndexOf() returned -1 saying expression evaluates to -1 and not a number. If I made it return >= 0 the condition evaluated fine. If I did a Condition with hardcoded -1 >= 0 it worked just fine. So a constant of -1 and a return value of >= 0 is a number, but a return value of -1 is not a number. Is this normal in msbuild?

**Reply by danmoseley** [reply]
You'd have to show the explicit examples, but generally in conditions MSBuild follows a series of heuristics to coerce argument types. eg
https://github.com/microsoft/msbuild/blob/4b437b2406d46ecc634a0b2c2ebc62254b0121f2/src/Build/Evaluation/Expander.cs#L4231

The result of IndexOf is as you say a number, so without the full example I can't say why it's triggering the "instead of a number" error here
https://github.com/microsoft/msbuild/blob/e70a3159d64f9ed6ec3b60253ef863fa883a99b1/src/Build/Evaluation/Conditionals/NumericComparisonExpressionNode.cs#L50

**Reply by maloo** [reply]
@danmosemsft 
I think I might have found the issue. For some odd reason MSBuild doesn't run in invariant mode and so my Swedish regional settings probably mess something up in the parsing of lists or numbers (we have different list separator ; and decimal point ,). This gets me much further, so maybe you need to add a regression test for building with Swedish regional settings ;)
I set DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 which gets me past the errors above. But I still get many errors while building. Maybe you can fix them all when fixing msbuild to not be culture dependent by default.
BTW. It was not my machine. I bought a new computer to test this and only installed dotnet, VS2019, SQL, BC, 7z, Git, N++ ;)

**Reply by maloo** [reply]
I will let you guys file the issue, not sure what repo it belong in, msbuild, runtime, sdk ...

**Reply by maloo** [reply]
And switching DOTNET_SYSTEM_GLOBALIZATION_INVARIANT off I get the same error again.

**Reply by safern** [reply]
It seems like this should be in msbuild because they should probably try to parse numbers with Invariant culture.

**Reply by maloo** [reply]
@rainersigwald, can you create an issue for msbuild?

**Reply by rainersigwald** [reply]
@maloo I'm not sure I understand the problem. It sounds like you have a minimal repro that fails on a Swedish machine--can you share that?

**Reply by maloo** [reply]
@rainersigwald Just an educated guess. But I think it should be enough to clone runtime repo and build using Swedish (sv-se) regional settings. I did this on a brand new machine too, English Win10, but Swedish regional settings. All details are in these discussions.

**Reply by danmoseley** [reply]
Not immediately obvious where the issue is, MSBuild parses numbers into strings here
https://github.com/microsoft/msbuild/blob/e70a3159d64f9ed6ec3b60253ef863fa883a99b1/src/Shared/ConversionUtilities.cs#L99
`return Double.Parse(number, NumberStyles.AllowDecimalPoint | NumberStyles.AllowLeadingSign, CultureInfo.InvariantCulture.NumberFormat);`

**Reply by danmoseley** [reply]
I can repro. I"ll open an issue.

**Reply by danmoseley** [reply]
https://github.com/microsoft/msbuild/issues/5502

**Comment by maloo** [comment]
Also tried a clean ~/.nuget to force update of all nugets. How can I check what msbuild get used? I have seen wrong versions being picked up as default sometimes. I have no msbuild in PATH.

**Reply by safern** [reply]
It should pick the one that comes with the dotnet CLI that we restore at the beginning of the build.

Do you have a `.dotnet` directory created in your repo?

**Reply by maloo** [reply]
Yes, the dotnet seems to be:

```
.NET SDK (reflecting any global.json):
 Version:   5.0.100-preview.6.20310.4
 Commit:    16d29b9349
```

**Reply by maloo** [reply]
Looks like I got 2 SDKs, but probably becuase I tried to build preview5 and master.
```
5.0.100-preview.5.20251.2
5.0.100-preview.6.20310.4
```

**Reply by maloo** [reply]
I hardcoded the expression IndexOf() >= 0 to False, and then I got a bit further. But it failed at a later stage in clr build at "win".SubString(0, -1), where -1 comes from RDI ("win") having no "-" so substring fails since length can't be -1. But at this point I could start building -subset libs, which also failed. And then I could open and build the library I needed using VS .sln
I really hope someone can figure out what is wrong. Shouldn't have to go through this to build a library ...

**Reply by eerhardt** [reply]
Can you dump out your environment variables? It sounds like your machine is in a weird state. Can you try a different machine?

**Reply by maloo** [reply]
```
ALLUSERSPROFILE=C:\ProgramData
APPDATA=C:\Users\marcu\AppData\Roaming
asl.log=Destination=file
CommonProgramFiles=C:\Program Files\Common Files
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
CommonProgramW6432=C:\Program Files\Common Files
COMPUTERNAME=XPS2MAN
ComSpec=C:\WINDOWS\system32\cmd.exe
DriverData=C:\Windows\System32\Drivers\DriverData
FSHARPINSTALLDIR=C:\Program Files (x86)\Microsoft SDKs\F#\10.1\Framework\v4.0\
HOMEDRIVE=C:
HOMEPATH=\Users\marcu
LOCALAPPDATA=C:\Users\marcu\AppData\Local
LOGONSERVER=\\XPS2MAN
NUMBER_OF_PROCESSORS=8
OneDrive=C:\Users\marcu\OneDrive
OneDriveConsumer=C:\Users\marcu\OneDrive
OS=Windows_NT
Path=C:\PROGRAM FILES\DELL\DW WLAN CARD;C:\PROGRAM FILES (X86)\INTEL\ICLS CLIENT\;C:\PROGRAM FILES\INTEL\ICLS CLIENT\;C:\WINDOWS\SYSTEM32;C:\WINDOWS;C:\WINDOWS\SYSTEM32\WBEM;C:\WINDOWS\SYSTEM32\WINDOWSPOWERSHELL\V1.0\;C:\PROGRAM FILES (X86)\INTEL\INTEL(R) MANAGEMENT ENGINE COMPONENTS\DAL;C:\PROGRAM FILES\INTEL\INTEL(R) MANAGEMENT ENGINE COMPONENTS\DAL;C:\PROGRAM FILES (X86)\INTEL\INTEL(R) MANAGEMENT ENGINE COMPONENTS\IPT;C:\PROGRAM FILES\INTEL\INTEL(R) MANAGEMENT ENGINE COMPONENTS\IPT;C:\PROGRAM FILES (X86)\NVIDIA CORPORATION\PHYSX\COMMON;C:\WINDOWS\SYSTEM32;C:\WINDOWS;C:\WINDOWS\SYSTEM32\WBEM;C:\WINDOWS\SYSTEM32\WINDOWSPOWERSHELL\V1.0\;C:\PROGRAM FILES\MICROSOFT SQL SERVER\CLIENT SDK\ODBC\110\TOOLS\BINN\;C:\PROGRAM FILES (X86)\MICROSOFT SQL SERVER\120\TOOLS\BINN\;C:\PROGRAM FILES\MICROSOFT SQL SERVER\120\TOOLS\BINN\;C:\PROGRAM FILES\MICROSOFT SQL SERVER\120\DTS\BINN\;C:\PROGRAM FILES (X86)\MICROSOFT SQL SERVER\120\TOOLS\BINN\MANAGEMENTSTUDIO\;C:\PROGRAM FILES (X86)\MICROSOFT SQL SERVER\120\DTS\BINN\;C:\PROGRAM FILES\GIT\CMD;C:\PROGRAM FILES\MICROSOFT\WEB PLATFORM INSTALLER\;C:\PROGRAM FILES\MICROSOFT SQL SERVER\130\TOOLS\BINN\;C:\Program Files\WIDCOMM\Bluetooth Software\;C:\Program Files\WIDCOMM\Bluetooth Software\syswow64;C:\Program Files (x86)\Microsoft SQL Server\Client SDK\ODBC\130\Tools\Binn\;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\Binn\;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn\;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\Binn\ManagementStudio\;C:\Program Files (x86)\Windows Live\Shared;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\Program Files\Microsoft Windows Performance Toolkit\;C:\Program Files (x86)\Microsoft SQL Server\140\Tools\Binn\;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn\;C:\Program Files (x86)\Microsoft SQL Server\140\Tools\Binn\ManagementStudio\;C:\Program Files\PuTTY\;C:\Program Files\nodejs\;C:\Program Files\TortoiseSVN\bin;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\dotnet\;C:\Program Files\Git\cmd;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files\Microsoft VS Code\bin;C:\Program Files\Docker\Docker\resources\bin;C:\ProgramData\DockerDesktop\version-bin;C:\Users\marcu\.dnx\runtimes\dnx-clr-win-x86.1.0.0-rc1-update1\bin;C:\Users\marcu\.dnx\bin;C:\Users\marcu\AppData\Local\Programs\Python\Python35\Scripts\;C:\Users\marcu\AppData\Local\Programs\Python\Python35\;C:\Users\marcu\AppData\Local\Microsoft\WindowsApps;C:\Users\marcu\AppData\Roaming\npm;C:\Users\marcu\.dotnet\tools;C:\Users\marcu\AppData\Local\Programs\Fiddler;C:\Users\marcu\AppData\Local\Microsoft\WindowsApps
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
PROCESSOR_ARCHITECTURE=AMD64
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 94 Stepping 3, GenuineIntel
PROCESSOR_LEVEL=6
PROCESSOR_REVISION=5e03
ProgramData=C:\ProgramData
ProgramFiles=C:\Program Files
ProgramFiles(x86)=C:\Program Files (x86)
ProgramW6432=C:\Program Files
PROMPT=$P$G
PSModulePath=C:\Users\marcu\Documents\WindowsPowerShell\Modules;C:\WINDOWS\system32\WindowsPowerShell\v1.0\Modules\;C:\Program Files (x86)\Microsoft SQL Server\120\Tools\PowerShell\Modules\;C:\Program Files (x86)\Microsoft SQL Server\130\Tools\PowerShell\Modules\;C:\Program Files\WindowsPowerShell\Modules\;C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\ResourceManager\AzureResourceManager\;C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\ServiceManagement\;C:\Program Files (x86)\Microsoft SDKs\Azure\PowerShell\Storage\
PUBLIC=C:\Users\Public
SystemDrive=C:
SystemRoot=C:\WINDOWS
TEMP=C:\Users\marcu\AppData\Local\Temp
TMP=C:\Users\marcu\AppData\Local\Temp
USERDOMAIN=XPS2MAN
USERDOMAIN_ROAMINGPROFILE=XPS2MAN
USERNAME=marcu
USERPROFILE=C:\Users\marcu
VS140COMNTOOLS=C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\Tools\
windir=C:\WINDOWS
WSLENV=WT_SESSION::WT_PROFILE_ID
WT_PROFILE_ID={61c64bbd-d2c6-5271-96e7-009a87fd44cf}
WT_SESSION=34e9fb34-2e2c-40f9-b6f2-9ad3941c7b79
```

**Reply by maloo** [reply]
Just one machine, WFH

**Reply by maloo** [reply]
Some debug info

```
PS C:\Data\Projects\GitHub\runtime2> .\build.cmd -restore -build -subset clr+libs -os Windows_NT -arch x64 -runtimeConfiguration Release -c Debug
win-x64
MSBuild C:\Users\marcu\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20256.5\tools\Build.proj   /p:Configuration=Debug /p:RepoRoot=C:\Data\Projects\GitHub\runtime2 /p:Restore=True /p:DeployDeps=False /p:Build=True /p:Rebuild=False /p:Deploy=False /p:Test=False /p:Pack=False /p:IntegrationTest=False /p:PerformanceTest=False /p:Sign=False /p:Publish=False /p:TargetArchitecture=x64 /p:subset=clr+libs /p:TargetOS=Windows_NT /p:RuntimeConfiguration=Release
```

**Reply by maloo** [reply]
I found some references to VSTA 14 and msbuild 14.0 on my machine and there is a msbuild 14 installed under Program files. I think this could be related to SQL server management tools. How can I find out what version/path of msbuild gets called by build.ps1? msbuild is not in the path when I start the build ...
When I run msbuild 14 in the runtime root folder it complains on the format of build.proj. When I run msbuild 15.1 (VS 2019 preview) it starts building everything, and so far no error. So maybe my feeling was right that an old/bad msbuild was picked up somehow.

**Reply by maloo** [reply]
no, removed all existing msbuilds in system, must be something else
